docs/slashing-protection.md

```diff
@@ -0,0 +1,105 @@
+# Slashing Protection on Finality Provider
+
+In the BTC staking protocol, finality providers run a
+Finality Provider Daemon (`fpd`) to send finality votes to Babylon.
+If a finality provider signs two conflicting blocks (same committed
+randomness but different block hash), all its delegations will be slashed.
+Apart from malicious behavior, honest finality providers face [slashing risks](https://cubist.dev/blog/slashing-risks-you-need-to-think-about-when-restaking)
+due to factors like hardware failures or software bugs.
+Therefore, a proper slashing protection mechanism is required.
+
+Recall that in our system, a finality provider needs to run two daemons:
+1. The finality provider daemon (fpd), which connects to the Babylon node
+   and initiates EOTS signing requests upon a new block to finalize
+2. The EOTS manager daemon (eotsd), which manages the EOTS key and responds to
+   signing requests from the finality provider daemon
+
+The two daemons have different responsibilities to prevent double-signing.
+The protections from the two daemons are complementary to each other.
+
+### Finality provider daemon protection
+
+**Assumption**:
+- The Babylon node the daemon connects to is trusted and responsive.
+- The `finality-provider.db` file is not compromised.
+
+The finality provider daemon ensures that it will never initiate
+a signing request for the same height twice if the previous request succeeds.
+
+To achieve this, the daemon does the following:
+
+- Maintains a local state `lastVotedHeight`, which is updated once
+  a vote submission succeeds, and never votes for a height that is not higher
+  than `lastVotedHeight`.
+- Polls blocks one-by-one in a monotonically increasing order.
+
+Once a finality provider is restarted, it needs to determine which height to
+start from, or bootstrapping. The bootstrapping needs to ensure that no blocks
+will be missed and voted blocks will not be polled again. The bootstrapping
+process is as follows:
+
+1. Query consumer chain for:
+   - `lastFinalizedHeight` (defaults to `0` if no blocks are finalized)
+   - `finalityActivationHeight`
+   - `highestVotedHeight` (defaults to `0` if no votes exist)
+
+2. If local state is empty or broken:
+   - Set `lastVotedHeight = lastFinalizedHeight`
+
+3. Compare `lastVotedHeight` and `highestVotedHeight`:
+   - If `lastVotedHeight >= highestVotedHeight`
+     - `startHeight = lastVotedHeight + 1`
+     - Note: this is possible if `highestVotedHeight` has not been updated due to
+       execution delay.
+   - If `lastVotedHeight < highestVotedHeight`
+     - Query consumer chain for whether the fp has voted for any blocks
+       in between `[lastVotedHeight + 1, highestVotedHeight)`
+     - Send votes if there are any missed blocks
+     - `startHeight = highestVotedHeight + 1`
+     - Note: this is possible due to bugs or if the local state is tampered with
```

Comment:
What do you mean by this being possible? Same with the sentence above.

Reply:
`lastVotedHeight` is maintained locally, while `highestVotedHeight` is from on-chain data. They should be the same value but could be otherwise due to execution latency or the local state is tampered with